### PR TITLE
Assignment2

### PR DESCRIPTION
**Pull request by kveselits**

### CODE REVIEW

**Comment by kveselits** [comment]
https://intellitect.visualstudio.com/EWU-CSCD/_build/results?buildId=7592

**Comment by CDobbins1** [comment]
Reviewing this

File: SecretSanta/src/SecretSanta.Data/ApplicationDbContext.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+using System.Threading.Tasks;
+
+namespace SecretSanta.Data
+{
+    public class ApplicationDbContext : DbContext
+    {
+        public DbSet<Gift>? Gifts { get; set; }
+        public DbSet<User>? Users { get; set; }
+        public DbSet<Group>? Groups { get; set; }
+        public DbSet<GroupSet>? GroupSets { get; set; }
```
Comment (CDobbins1) [inline SecretSanta/src/SecretSanta.Data/ApplicationDbContext.cs]: I am not sure if you need DbSet for GroupSets.

File: SecretSanta/test/SecretSanta.Data.Tests/GiftTests.cs [hunk 2]
```diff
@@ -5,13 +5,20 @@
...
             // Arrange
-            Gift gift = new Gift(1, "Ring 2", "Amazing way to keep the creepers away", "www.ring.com", new User(1, "Inigo", "Montoya", new List<Gift>()));
+            Gift gift = new Gift
+            {
+                Id = 1,
+                Title = "Ring 2",
+                Description = "Amazing way to keep the creepers away",
+                Url = "www.ring.com",
+                User = new User {Id = 1, FirstName = "Inigo", LastName = "Montoya", Gifts = new List<Gift>()}
+            };
```
Comment (CDobbins1) [inline SecretSanta/test/SecretSanta.Data.Tests/GiftTests.cs]: Stokes mentioned in gitter that it is okay to remove the preexisting tests that were meant for testing the constructors.

File: SecretSanta/test/SecretSanta.Data.Tests/GiftTests.cs [hunk 3]
```diff
@@ -27,21 +34,47 @@ public void Gift_CanBeCreate_AllPropertiesGetSet()
...
+                Description = "Amazing way to keep the creepers away",
+                Url = null!,
+                User = new User {Id = 1, FirstName = "Inigo", LastName = "Montoya", Gifts = new List<Gift>()}
+            };
+        }
+
+        [TestMethod]
+        public void Gift_AddGift_ShouldCreateForeignRelationship()
+        {
         }
```
Comment (CDobbins1) [inline SecretSanta/test/SecretSanta.Data.Tests/GiftTests.cs]: Looks like you forgot about this test?

Comment (kveselits) [reply SecretSanta/test/SecretSanta.Data.Tests/GiftTests.cs]: Yeah, I didn't give myself enough time to finish the assignment so there are a lot of things I knew I needed to test but ran out of time.

**Comment by breaman** [comment]
##  (kveselits)

## Assignment2

| Project Analysis Summary |   |
| -----------------| - |
| Nullable Enabled: | **✔** |
| Unexpected Disabled Warnings:  | **✔**
| Appropriately handle all warnings: | **✔** |

| ProjectName | Nullable | Solutions | Analyzers | Unexpected Disabled Warnings | Waning Level |
| -- | -- | -- | -- | -- | -- |
| SecretSanta.Api | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Business | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Data | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Web | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer |  | 4 |
| SecretSanta.Api.Tests | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Business.Tests | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Data.Tests | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Web.Tests | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer |  | 4 |


## Main Assignment

1. Add the following classes to the **SecretSanta.Data** project with the following associated properties:
   - **`EntityBase`** ✔
     - `int Id`
   - **`FingerPrintEntityBase`** inherit from **`EntityBase`** ✔
     - `string CreatedBy`
     - `DateTime CreatedOn`
     - `string ModifiedBy`
     - `DateTime ModifiedOn`
   - **`Group`** inherit from **`FingerPrintEntityBase`** ✔
     - `string Name`
   - Create a class that will hold the many to many relationship between User and Group ✔
     - Add a collection on both User and Group that references this class
2. Modify the following classes in the **SecretSanta.Data** project:
   - **`Gift`** ✔
     - Remove the Id property and inherit from FingerPrintEntityBase
   - **`User`** ✔
     - Remove the Id property and inherit from FingerPrintEntityBase
     - Make `Gifts` read/write
     - Create a nullable Santa property that contains a User as someone's Santa
3. Remove non-default constructors for `Gift` and `User`. ✔
4. In the **SecretSanta.Data** project
   - Create the **`ApplicationDbContext`** class ✔
     - Create two constructors ✔
        - One only takes the DbContextOptions and passes it on to the base class
        - One that takes the DbContextOptions and an IHttpContextAccessor, pass the DbContextOptions to the base class and save off the IHttpContextAccessor in a class property
     - Add DbSet properties for User, Gift, and Group classes ✔
     - Override OnModelCreating method and use the fluent API to configure the many-to-many pieces that are needed for setting up the database properly ✔
     - Override both SaveChanges methods and apply the fingerprinting logic ✔
5. Update unit tests in the **`SecretSanta.Data.Tests`** project to test actual database logic instead of just constructor logic
   - Mock up IHttpContextAccessor to verify that fingerprint information is working correctly (using Moq library) ❌


## Extra Credit

The following are options for extra credit:
- Enable console logging that will log all database commands ✔

**Pull request closed without merging**